barretenberg/cpp/pil/vm2/keccakf1600.pil

```diff
@@ -1,8 +1,1402 @@
+include "bitwise.pil";
+include "constants_gen.pil";
+include "range_check.pil";
+include "precomputed.pil";
+include "keccak_memory.pil";
+
 namespace keccakf1600;
 
-// Dummy.
-pol commit keccak[1000];
+// Reference document: https://keccak.team/files/Keccak-reference-3.0.pdf
+// Overview: https://keccak.team/keccak_specs_summary.html
+//
+// A keccakf1600 permutation is the result of 24 (AVM_KECCAKF1600_NUM_ROUNDS) iterations
+// of a round function.
+// The round function is composed of 5 sub-functions: theta, rho, pi, chi, and iota.
+// Main low-level operations are: XOR, AND, bit rotations
+// All the sub-functions operate over a state of 25 64-bit values.
+
+//#################################################################################################
+//                  High-Level Trace Structure Columns
+//#################################################################################################
+//   ctr   sel   start   last    state_in_00  state_in_01 ....
+//    1     1      1      0
+//    2     1      0      0         ROUND FUNCTION SPECIFICS
+//    3     1      0      0
+//    ...  ...    ...    ...               ...
+//    24    1      0      1
+
+// We use one row to compute a single round function.
+// 24 rows correspond to one keccak permutation.
+// Only on last row, we constrain the output columns.
+
+// keccakf1600 selector
+pol commit sel;
+sel * (1 - sel) = 0;
+
+// No relations will be checked if this identity is satisfied.
+#[skippable_if]
+sel + last = 0;
+
+// error is defined below after round function
+pol commit sel_no_error;
+#[SEL_NO_ERROR]
+start * (sel_no_error - 1 + error) = 0;
+
+// Round number which starts at 1 and ends at AVM_KECCAKF1600_NUM_ROUNDS (24).
+pol commit round;
+
+// First round of a keccak permutation. This is where the multi-rows computation starts and selector start
+// is used by the caller to trigger the keccak permutation computation.
+pol commit start;
+start * (1 - start) = 0;
+
+// We enforce the initial round to be set to 1.
+start * (round - 1) = 0;
+
+// sel == 1 <==> round != 0
+pol commit round_inv;
+#[KECCAK_SEL_ROUND_NON_ZERO]
+round * ((1 - sel) * (1 - round_inv) + round_inv) - sel = 0;
+
+#[KECCAK_ROUND_INCREMENT]
+// Note: sel factor is required for an empty row to satisfy this relation
+sel * (1 - last) * (round' - round - 1) = 0;
+
+// Last round of a keccak permutation
+pol commit last;
+last * (1 - last) = 0;
+
+// We define last == 1 <==> round == AVM_KECCAKF1600_NUM_ROUNDS (24)
+pol commit round_min_num_rounds_inv;
+#[KECCAK_LAST]
+sel * ((round - constants.AVM_KECCAKF1600_NUM_ROUNDS) * (last * (1 - round_min_num_rounds_inv) + round_min_num_rounds_inv) + last - 1) = 0;
```

Comment:
Something else for consideration (didn't think a lot about it): if you _do_ enforce that `last` is set in the 25th round...

You might be able to avoid needing this constraint if you put a fixed `25` in the lookup to the slice write.

```
#[WRITE_TO_SLICE]
sel_slice_write {...
      clk, dst_addr, space_id
      round}
is
keccak_memory.start_write {...
                           keccak_memory.clk, keccak_memory.addr, keccak_memory.space_id,
                           25};
```

if there is some error, this doesn't matter, and you are also free to set `last` as you wish (in particular, on the first row?). If there is no error, then `sel_slice_write` has to happen with `last` (you have a constrain for this). Then this means that
`last <=> sel_slice_write <=> round = 25`

Reply:
Another way to see it is that you'd be "batching" the inv check of `round = 25` into the inverse check of the lookup itself 😅